Restructure multiline and grouped bar chart functions

MelinaGoula opened the pull request:
Functions to review:
- `horizontal_grouped_bar_chart `
- `vertical_grouped_bar_chart`
- `multi_linechart_test`
- helper functions (lines 1102 to 1490)

MelinaGoula commented:
**Comparison of stacked bar chart functions** 

Previous (`stacked_chart`)
![sphx_glr_plot_grouped_bar_rick_001](https://github.com/CityofToronto/bdit_plotting_gallery/assets/107000709/a21fb507-1b04-48e3-9de6-426f7d6792a6)
Rewrite (`horizontal_grouped_bar_chart`) 
![sphx_glr_plot_grouped_bar_rick_002](https://github.com/CityofToronto/bdit_plotting_gallery/assets/107000709/c5acae87-3b85-4aa2-8353-8f3382fff92e)
Example of (`vertical_grouped_bar_chart`) 
![sphx_glr_plot_grouped_bar_rick_003](https://github.com/CityofToronto/bdit_plotting_gallery/assets/107000709/852ba64f-0954-4f00-bef6-b5a338d3ddaf)

MelinaGoula commented:
Functionality that was not carried over from `multi_linechart` to `multi_linechart_test` (the temporary new version): 
- Ability to set custom styling parameters (font-size, font-family, fontfamily-list, stroke, stroke-width, border)
- Setting a plot title
- Dashed grid lines as opposed to solid grid lines 
- Customizable tick length and width 
- Customizable sizes for axis and tick labels 
- Customizable precision for tick labels 
- Ability to set a specific line colour, width and border style 
- Option to add shaded areas 

Example plot with `multi_linechart` showing a lot of the functionality 
![sphx_glr_plot_multiline_003](https://github.com/CityofToronto/bdit_plotting_gallery/assets/107000709/a4ee8559-b0eb-4565-831c-1dc82ab0225d)

Recreated with `multi_linechart_test`: 
![sphx_glr_plot_multiline_002](https://github.com/CityofToronto/bdit_plotting_gallery/assets/107000709/a85466f0-fda7-46b1-983d-2c67ca25f31e)
I am not sure how to deal with legend placement in cases like this where there is not a lot of free space. One alternative is to follow the previous format of having multiple columns in the legend which would look like this: 
![sphx_glr_plot_multiline_002](https://github.com/CityofToronto/bdit_plotting_gallery/assets/107000709/866c7c89-0eea-4a27-a40c-1da7b181cfb1)

gabrielwol commented on rick.py:
```diff
@@ -598,3 +855,901 @@ def bar_chart(data_in, xlab,**kwargs):
+    
+        delta = (ymax - ymin)/4
+        i = 0
+        while True:
```
Should this be using `calculate_delta`? Same comment on line 580.

gabrielwol commented on rick.py:
```diff
@@ -598,3 +855,901 @@ def bar_chart(data_in, xlab,**kwargs):
+        ax = plt.gca()
+    
+        # ----------------------------------------------
+        # Default styling params if not defined in sett
+        if 'body' in sett:
+            dflt={
+                'font-size':(12 if 'font-size' not in sett['body']
+                             else sett['body']['font-size']),
+                'font-family':('sans-serif' if 'font-family' not in sett['body']
+                             else sett['body']['font-family']),
+                'fontfamily-list':(['Libre Franklin', 'DejaVu Sans'] if 'fontfamily-list'
+                              not in sett['body']
+                              else sett['body']['fontfamily-list']),
+                'stroke':('#000000' if 'stroke' not in sett['body']
+                             else sett['body']['stroke']),
+                'stroke-width':(2 if 'stroke-width' not in sett['body']
+                             else sett['body']['stroke-width']),
+                'border':('solid' if 'border' not in sett['body']
+                             else sett['body']['border'])
+            }
+        else:
+            dflt={
+                'font-size':12, 'font-family':'sans-serif',
+                'fontfamily-list':['Libre Franklin', 'DejaVu Sans'],
+                'stroke':'#000000', 'stroke-width':2, 'border':'solid'
+            }
```
Should default stylings live outside of a single graph function?

gabrielwol commented:
> I am not sure how to deal with legend placement in cases like this where there is not a lot of free space. One alternative is to follow the previous format of having multiple columns in the legend which would look like this:

Have you considered having the legend always **outside** the plot? This could help with consistency no matter the data being plotted!

gabrielwol commented:
> Functionality that was not carried over from multi_linechart to multi_linechart_test (the temporary new version):
....
Recreated with multi_linechart_test:
![302367434-a85466f0-fda7-46b1-983d-2c67ca25f31e](https://github.com/CityofToronto/bdit_plotting_gallery/assets/80077912/0bdb5a86-3aec-4c27-b8c7-07e549c67730)

I think the ticks would be an important feature for this graph type which we often use for timeseries. In our demos people are always trying to pinpoint the specific date that graphed events happen. Maybe dark ticks for the labelled points and lighter ticks for the others as a default?

gabrielwol commented:
@MelinaGoula your 3 new charting functions are very elegant compared with the old versions, well done! See a few comments above.

MelinaGoula commented on rick.py:
```diff
@@ -598,3 +855,901 @@ def bar_chart(data_in, xlab,**kwargs):
+def calculate_delta(max_value:float, min_value:float) -> (float, float):
+    '''
+    Returns parameters used to find the size of the y or x axis increments.
+    '''
```
Added!

MelinaGoula commented:
I have made the following changes and I think this is ready for re-review:
- added the option to add minor ticks to the multi line chart function and specify their spacing, e.g.
![image](https://github.com/CityofToronto/bdit_plotting_gallery/assets/107000709/63d0c317-d5f7-440c-a779-de41527b0b33)
- moved the legend to the outside of plots 
![image](https://github.com/CityofToronto/bdit_plotting_gallery/assets/107000709/8f81e566-19a8-46b2-a54d-2c7388d2e310)
- added the option for labeled shaded areas, e.g. 
![image](https://github.com/CityofToronto/bdit_plotting_gallery/assets/107000709/af1c4186-6fef-42f3-9697-0145520ff8d7)

One thing I think can be done in the future to improve the code structure is to not have as many function parameters passed in the helper functions, but this can be probably only be done by refactoring the code.

gabrielwol reviewed (approved):
These changes look great! Well done @MelinaGoula

radumas reviewed (changes requested):
Got a slight misfire in building Sphinx. 

We need the `docs/.nojekyll` file to remain.

[Also looks like we don't need to commit `.doctree` files (can add this to .gitignore) ](https://stackoverflow.com/a/33960721/4047679)

The contents of the `html` folder should be in `docs`.

Maybe we should change the build config back to building to `sphinx/build` and have instructions for people to manually copy the contents of `html` over to `docs` when they are ready to merge

radumas reviewed (approved):
Lots of changes here but I think this is really going to help clean up the repo!
